### PR TITLE
Parse reST on background thread and save scroll position

### PR DESCRIPTION
**Pull request by JanKanis**
Hi Peter, I made some changes to the reST plugin. 

I wanted to use it with a large-ish reST document (about 1300 lines), and this plugin wasn't very usable because after typing one character it would take close to a second to refresh the preview, freezing gedit in the mean time. The primary change in this PR is to move the reST parsing to a background thread. I also tried doing it in an `idle_add` handler, but that didn't help enough. After that I also made some other changes that were useful with large documents.

- Parse reST document in a background thread
- Save the scroll position between redraws of the preview, and also when switching to non-reST files and back.
- Only refresh if the preview panel is visible
- Don't load webkit until we need it, allowing the plugin to be enabled when not actively in use.

I prefer to have the preview in the side panel, but—not having much GTK experience—I didn't venture into making a preferences panel for the plugin, so at the moment I'm just keeping that as a local hardcoded change. I did make some changes so that there's only one place where `window.get_bottom_panel()` gets called, so it can be changed easily.

Please let me know if you need any changes for this PR to be merged.

### CODE REVIEW

**Comment by bittner** [comment]
Hey, cool! This is really appreciated! I'll take a look at your changes as soon as I can spare a decent amount of time.

Can you share details about the environment you developed and tried this out in? (GNU/Linux distro, GNOME version, etc.)

**Comment by JanKanis** [comment]
Sure, I made this on Ubuntu 18.04. Gedit 3.28.1, Gnome 3.22, Python 3.6.9. There is a small bit of the code linux specific (lowering the background thread priority), but that should degrade gracefully.

**Review by bittner — CHANGES_REQUESTED** [review]
That seems to work nicely, congratulations!

I only noticed that, when you select text, it doesn't update the end position of the caret when you move the selection any further. Only the start and the very first end caret position (when you stop the first time during selecting text) is taken for the rendering. I'll try to find out how that can be fixed.

For the rest of the enhancements, I made a few comments.

If you don't mind, please run `flake8` and, optionally, `pylint` over the files you changed. I'd like to add linting later, over the entire code base, so we can start taking control back over the code. :grimacing: :see_no_evil:

File: reST/restructuredtext.py [hunk 1]
```diff
@@ -23,8 +23,27 @@
...
+class State(Enum):
+    NON_REST = 1
+    REST = 2
+    SELECTION = 3
+    EXIT = 4
+
+DEBUG = False
+def debug(*args):
+    if DEBUG:
+        print(*args, file=sys.stderr)
```
[the remark this answers is not in the document]
Comment (bittner) [reply reST/restructuredtext.py:41]: There even seems to be a [gedit_debug](https://developer.gnome.org/gedit/stable/gedit-gedit-debug.html) function. But I have no idea what it does, precisely, in the end.

[the remark this answers is not in the document]
Comment (JanKanis) [reply reST/restructuredtext.py:41]: I added the debug output to debug the state changes that are happening concurrently. I think it's mostly useful if someone is modifying the code, in which case changing it to `DEBUG = True` is trivial. It is really more a 'trace' level of output. So I thought just a simple constant flag was good enough for this. Python logging would also work but I don't see the benefit for this case.

File: reST/restructuredtext.py [hunk 2]
```diff
@@ -42,57 +61,190 @@ class RestructuredtextHtmlPanel(Gtk.ScrolledWindow):
...
+
+    def rest_parser_thread(self):
+        try:
+            tid = int(os.readlink('/proc/thread-self').split('/')[-1])
+            debug("reST preview rendering thread id", tid)
+            # Set nice +10 and SCHED_BATCH on this thread
+            os.system(f'schedtool -n 10 -B {tid}')
+        except OSError as e:
+            debug("Unable to set batch priority for reST preview rendering thread. "
+                  "This is only supposed to work on Linux")
```
[the remark this answers is not in the document]
Comment (JanKanis) [reply reST/restructuredtext.py:231]: I think the readme file is actually the better place to document this. No regular user is going to read terminal output from a graphical text editor, regular users start gedit by clicking the icon in which case the standard outputs are lost. But I should update this to print a message not through `debug`.

**Comment by bittner** [comment]
> I prefer to have the preview in the side panel, but—not having much GTK experience—I didn't venture into making a preferences panel for the plugin

I have prepared a Preferences dialog, which we could integrate with your changes. Can I push to this PR, or attach a patch file, or do you prefer that I integrate this after merging your changes?

File: reST/restructuredtext.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bittner) [reply reST/restructuredtext.py:41]: The logging module is optimized for being left in the code. It's designed to not add performance burden when executed, and it's "trivial" to configure output targets and change formatting, centrally. 

The latter criteria are covered with your implementation too, granted. But `print` statements can be problematic, especially for desktop applications, hence they should be avoided. And why not do it "the right way"?

[the remark this answers is not in the document]
Comment (bittner) [reply reST/restructuredtext.py:41]: It looks like that with [gedit_debug](https://developer.gnome.org/gedit/stable/gedit-gedit-debug.html) you can [turn on debug output](https://stackoverflow.com/questions/13790559/showing-gtk-debug-log-messages) via an environment variable, no need to change the code. More details about are covered by the docs at [Running and debugging GTK+ Applications](https://developer.gnome.org/gtk3/stable/gtk-running.html). This approach seems to have additional benefits, as e.g. the "Interactive debugging" section in the same document demonstrates.

**Comment by JanKanis** [comment]
> > I prefer to have the preview in the side panel, but—not having much GTK experience—I didn't venture into making a preferences panel for the plugin
> 
> I have prepared a Preferences dialog, which we could integrate with your changes. Can I push to this PR, or attach a patch file, or do you prefer that I integrate this after merging your changes?

I don't think you can push to this PR as the lives in my github account, but you can push to a branch in your own repository, then I can pull it.

File: reST/restructuredtext.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JanKanis) [reply reST/restructuredtext.py:41]: The gedit_debug functionality doesn't look like a good match for Python code (it requires manual file/line numbers). If you prefer using python logging I'm fine with that, I'll update it.

**Comment by bittner** [comment]
> > > I prefer to have the preview in the side panel, but—not having much GTK experience—I didn't venture into making a preferences panel for the plugin
> > 
> > 
> > I have prepared a Preferences dialog, which we could integrate with your changes. Can I push to this PR, or attach a patch file, or do you prefer that I integrate this after merging your changes?
> 
> I don't think you can push to this PR as the lives in my github account, but you can push to a branch in your own repository, then I can pull it.

I have pushed one commit to the [pr/16](https://github.com/bittner/gedit-reST-plugin/commits/pr/16) branch. It is the dialog as such. I still need to figure out how to persist and read out the settings reliably with `Gio.Settings`. Seems like a tricky topic.

File: reST/restructuredtext.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bittner) [reply reST/restructuredtext.py:41]: Agreed. If you would add Python logging that would be helpful. Thank you! :+1:

**Comment by JanKanis** [comment]
> > > > I prefer to have the preview in the side panel, but—not having much GTK experience—I didn't venture into making a preferences panel for the plugin
> > > 
> > > 
> > > I have prepared a Preferences dialog, which we could integrate with your changes. Can I push to this PR, or attach a patch file, or do you prefer that I integrate this after merging your changes?
> > 
> > 
> > I don't think you can push to this PR as the lives in my github account, but you can push to a branch in your own repository, then I can pull it.
> 
> I have pushed one commit to the [pr/16](https://github.com/bittner/gedit-reST-plugin/commits/pr/16) branch. It is the dialog as such. I still need to figure out how to persist and read out the settings reliably with `Gio.Settings`. Seems like a tricky topic.

I actually also started having a look into a prefs dialog and saving settings. The instructions [here](https://askubuntu.com/questions/251712/how-can-i-install-a-gsettings-schema-without-root-privileges) looked like the best way to go, storing the compiled schema in the reST directory.

**Comment by bittner** [comment]
> The instructions [here](https://askubuntu.com/questions/251712/how-can-i-install-a-gsettings-schema-without-root-privileges) looked like the best way to go, storing the compiled schema in the reST directory.

Sounds good.

Do you want me to put that into code, or are you planning to implement that yourself? Let me know!

**Comment by JanKanis** [comment]
> > The instructions [here](https://askubuntu.com/questions/251712/how-can-i-install-a-gsettings-schema-without-root-privileges) looked like the best way to go, storing the compiled schema in the reST directory.
> 
> Sounds good.
> 
> Do you want me to put that into code, or are you planning to implement that yourself? Let me know!

I don't have time to work on this the coming few days, maybe again in the weekend. So if you want to pick it up that's fine, if not I'll continue with it when I have time to work on this again.

**Comment by JanKanis** [comment]
> I only noticed that, when you select text, it doesn't update the end position of the caret when you move the selection any further. Only the start and the very first end caret position (when you stop the first time during selecting text) is taken for the rendering. I'll try to find out how that can be fixed.

I noticed that too, but it was already broken like that when I started with this. It needs to receive a notification when the selection changes, but I haven't looked into where to get that.

**Review by bittner — APPROVED** [review]
Let fix the debug logging, the preferences dialog and aesthetic changes as a follow-up PR.

**Comment by bittner** [comment]
Thank you for this awesome contribution! :1st_place_medal:

**Pull request merged**